In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
import os

# Explore data



In [ ]:
crawling_all= pd.read_csv('/content/drive/MyDrive/WiWi/Crawling_all.csv', parse_dates=['Last Modified'])

<ipython-input-3-e03dea8bfb0e>:1: DtypeWarning: Columns (19,20) have mixed types. Specify dtype option on import or set low_memory=False.
  crawling_all= pd.read_csv('/content/drive/MyDrive/WiWi/Crawling_all.csv', parse_dates=['Last Modified'])


In [ ]:
crawling_all.shape

(1487191, 88)

In [ ]:
file_size=os.path.getsize("/content/drive/MyDrive/WiWi/Crawling_all.csv")

In [ ]:
size = file_size / 1024 ** 3
print(size)

1.3837974593043327


In [ ]:
crawling_all.head()

Address  \
0  http://www.fsjpst.rnu.tn/fr/mastere-de-gouvern...   
1  http://www.utm.rnu.tn/visirech/Fr/utm/fsegt/if...   
2  https://www.justice.gov.tn/index.php?id=192&tx...   
3  http://opendata.interieur.gov.tn/ar/datas/inde...   
4  http://opendata.interieur.gov.tn/fr/outils/ind...   

               Content Type  Status Code Status Indexability  \
0  text/html; charset=UTF-8          200     OK    Indexable   
1  text/html; charset=UTF-8          200     OK    Indexable   
2   text/html;charset=utf-8          200     OK    Indexable   
3  text/html; charset=UTF-8          200     OK    Indexable   
4  text/html; charset=UTF-8          200     OK    Indexable   

  Indexability Status                                 Title 1  Title 1 Length  \
0                 NaN                                     NaN               0   
1                 NaN                                VISIRECH               8   
2                 NaN              وزارة العدل : طلبات العروض              26   
3                 NaN  موقع البيانات المفتوحة لوزارة الداخلية              38   
4                 NaN                                   قائمة               5   

   Title 1 Pixel Width                                 Meta Description 1  \
0                    0                                                NaN   
1                   85  Plateforme de la recherche scientifique de l'U...   
2                  214  ﻭﺯﺍﺭﺓ ﺍﻟــــــﻌـــــــﺪﻝ ﺍﻟﺠـــــﻤﻬـــﻮﺭﻳﺔ ﺍﻟﺘ...   
3                  326  Le site officiel des données ouvertes du Minis...   
4                   45  Le site officiel des données ouvertes du Minis...   

   ...  Mobile Usability  Mobile Usability Issues AMP Results  AMP Issues  \
0  ...               NaN                      NaN         NaN         NaN   
1  ...               NaN                      NaN         NaN         NaN   
2  ...               NaN                      NaN         NaN         NaN   
3  ...               NaN                      NaN         NaN         NaN   
4  ...               NaN                      NaN         NaN         NaN   

  Rich Results  Rich Results Types Rich Results Type Errors  \
0          NaN                 NaN                      NaN   
1          NaN                 NaN                      NaN   
2          NaN                 NaN                      NaN   
3          NaN                 NaN                      NaN   
4          NaN                 NaN                      NaN   

   Rich Results Type Warnings  \
0                         NaN   
1                         NaN   
2                         NaN   
3                         NaN   
4                         NaN   

                                 URL Encoded Address      Crawl Timestamp  
0  http://www.fsjpst.rnu.tn/fr/mastere-de-gouvern...  2023-07-29 17:51:47  
1  http://www.utm.rnu.tn/visirech/Fr/utm/fsegt/if...  2023-07-27 17:11:19  
2  https://www.justice.gov.tn/index.php?id=192&tx...  2023-07-22 20:25:29  
3  http://opendata.interieur.gov.tn/ar/datas/inde...  2023-07-24 22:17:58  
4  http://opendata.interieur.gov.tn/fr/outils/ind...  2023-07-27 21:55:58  

[5 rows x 88 columns]

In [ ]:
crawling_all.columns

Index(['Address', 'Content Type', 'Status Code', 'Status', 'Indexability',
       'Indexability Status', 'Title 1', 'Title 1 Length',
       'Title 1 Pixel Width', 'Meta Description 1',
       'Meta Description 1 Length', 'Meta Description 1 Pixel Width',
       'Meta Keywords 1', 'Meta Keywords 1 Length', 'H1-1', 'H1-1 Length',
       'H2-1', 'H2-1 Length', 'Meta Robots 1', 'X-Robots-Tag 1',
       'Meta Refresh 1', 'Canonical Link Element 1', 'rel="next" 1',
       'rel="prev" 1', 'HTTP rel="next" 1', 'HTTP rel="prev" 1',
       'amphtml Link Element', 'Size (bytes)', 'Word Count', 'Sentence Count',
       'Average Words Per Sentence', 'Flesch Reading Ease Score',
       'Readability', 'Text Ratio', 'Crawl Depth', 'Link Score', 'Inlinks',
       'Unique Inlinks', 'Unique JS Inlinks', '% of Total', 'Outlinks',
       'Unique Outlinks', 'Unique JS Outlinks', 'External Outlinks',
       'Unique External Outlinks', 'Unique External JS Outlinks',
       'Closest Similarity Match', 'No. Ne

# Check columns

In [ ]:
#Check columns with 100% empty cells
for col in crawling_all.columns:
    if crawling_all[col].isnull().all():
        print(col)

Canonical Link Element 1
rel="next" 1
rel="prev" 1
HTTP rel="next" 1
HTTP rel="prev" 1
amphtml Link Element
Link Score
Inlinks
Unique Inlinks
Unique JS Inlinks
% of Total
Outlinks
Unique Outlinks
Unique JS Outlinks
External Outlinks
Unique External Outlinks
Unique External JS Outlinks
Closest Similarity Match
No. Near Duplicates
Spelling Errors
Grammar Errors
Clicks
Impressions
CTR
Position
URL Inspection API Status
URL Inspection API Error
Summary
Coverage
Last Crawl
Days Since Last Crawled
Crawled As
Crawl Allowed
Page Fetch
Indexing Allowed
User-Declared Canonical
Google-Selected Canonical
Mobile Usability
Mobile Usability Issues
AMP Results
AMP Issues
Rich Results
Rich Results Types
Rich Results Type Errors
Rich Results Type Warnings


In [ ]:
# Drop columns with empty cells
crawling_all.dropna(how='all', axis=1, inplace=True)

In [ ]:
crawling_all.shape

(1487191, 43)

# Create a copy from crawling_all


In [ ]:
df = crawling_all.copy()

In [ ]:
df.columns

Index(['Address', 'Content Type', 'Status Code', 'Status', 'Indexability',
       'Indexability Status', 'Title 1', 'Title 1 Length',
       'Title 1 Pixel Width', 'Meta Description 1',
       'Meta Description 1 Length', 'Meta Description 1 Pixel Width',
       'Meta Keywords 1', 'Meta Keywords 1 Length', 'H1-1', 'H1-1 Length',
       'H2-1', 'H2-1 Length', 'Meta Robots 1', 'X-Robots-Tag 1',
       'Meta Refresh 1', 'Size (bytes)', 'Word Count', 'Sentence Count',
       'Average Words Per Sentence', 'Flesch Reading Ease Score',
       'Readability', 'Text Ratio', 'Crawl Depth', 'Hash', 'Response Time',
       'Last Modified', 'Redirect URL', 'Redirect Type', 'Cookies',
       'PDF Subject', 'PDF Author', 'PDF Created', 'PDF Modified',
       'PDF Page Count', 'HTTP Version', 'URL Encoded Address',
       'Crawl Timestamp'],
      dtype='object')

# Take only rows with Content Type = PDF or HTML

In [ ]:
df.shape

(1487191, 43)

In [ ]:
df = df[(df['Content Type'].str.contains('HTML', case=False)) | (df['Content Type'].str.contains('PDF', case=False)) ]

In [ ]:
df.shape

(1381376, 43)

In [ ]:
df['Content Type'].value_counts()

text/html; charset=UTF-8                                                                                                                                                                                                                                                  1035785
text/html; charset=utf-8                                                                                                                                                                                                                                                   217593
application/pdf                                                                                                                                                                                                                                                             68399
text/html; charset=iso-8859-1                                                                                                                                                     

In [ ]:
# Replace 'html' and 'pdf' values in the 'Content Type' column
df.loc[df['Content Type'].str.contains('html', case=False), 'Content Type'] = 'html'
df.loc[df['Content Type'].str.contains('pdf', case=False), 'Content Type'] = 'pdf'

In [ ]:
df['Content Type'].value_counts()

html    1312956
pdf       68420
Name: Content Type, dtype: int64

# Take rows with Status Code = 200

In [ ]:
df['Status Code'].value_counts()

200    1315200
404      28722
302      16372
301      10270
503       4661
303       4574
500        682
403        519
401        202
307        172
400          2
Name: Status Code, dtype: int64

In [ ]:
df = df[df['Status Code'] == 200]

In [ ]:
df.shape

(1315200, 43)

# Check for duplicated hash values and remove duplicates

In [ ]:
df['Hash'].duplicated().value_counts()

False    1151803
True      163397
Name: Hash, dtype: int64

In [ ]:
# Drop rows with duplicated hash values and keep the first occurrence
mask = df.duplicated(subset='Hash', keep='first') & ~df['Hash'].isna()

# Filter the DataFrame using the mask
df = df[~mask]

In [ ]:
df['Hash'].duplicated().value_counts()

False    1151803
True       68747
Name: Hash, dtype: int64

In [ ]:
df[df['Content Type'].str.contains('pdf',case=False)].shape

(68420, 43)

In [ ]:
df.shape

(1220550, 43)

# Remove addresses with obvious irrelevant keywords

In [ ]:
df = df[~df['Address'].str.contains('login')]

In [ ]:
df = df[~df['Address'].str.contains('register',case=False)]

In [ ]:
df = df[~df['Address'].str.contains('user',case=False)]

In [ ]:
df = df[~df['Address'].str.contains('password',case=False)]

In [ ]:
df = df[~df['Address'].str.contains('contact',case=False)]

In [ ]:
df = df[~df['Address'].str.contains('profile',case=False)]

In [ ]:
# Remove addresses of campus france of other countries != tunisie
df = df[~(df['Address'].str.contains('campusfrance',case=False) & ~df['Address'].str.contains('tunisie',case=False))]

In [ ]:
df = df[~df['Address'].str.contains('agenda',case=False)]

In [ ]:
df = df[~df['Address'].str.contains('partenaire',case=False)]

In [ ]:
df = df[~df['Address'].str.contains('add-to-cart',case=False)]

In [ ]:
df = df[~df['Address'].str.contains('produit',case=False)]

In [ ]:
df = df[~df['Address'].str.contains('product',case=False)]

In [ ]:
df = df[~df['Address'].str.contains('نعي',case=False)]

In [ ]:
df = df[~df['Address'].str.contains('send',case=False)]

In [ ]:
df = df[~df['Address'].str.contains('mail',case=False)]

In [ ]:
df = df[~df['Address'].str.contains('photo',case=False)]

In [ ]:
df = df[~df['Address'].str.contains('video',case=False)]

In [ ]:
df = df[~df['Address'].str.contains('album',case=False)]

In [ ]:
df = df[~df['Address'].str.contains('fete',case=False)]

In [ ]:
df = df[~df['Address'].str.contains('lost_pwd',case=False)]

In [ ]:
df = df[~df['Address'].str.contains('authentif',case=False)]

In [ ]:
df = df[~df['Address'].str.contains('ميديا')]

In [ ]:
df.shape

(1132960, 43)

# Crawl Depth

In [ ]:
# Check rows with high Crawl Depth
high_depth = df[df['Crawl Depth']>1000][['Address','Content Type', 'Crawl Depth', 'Size (bytes)','Word Count', 'Sentence Count']].sort_values(by='Crawl Depth', ascending=False)
high_depth.shape

(34303, 6)

## Check addresses related to ONSR website

In [ ]:
high_depth[high_depth['Address'].str.contains('onsr', case=False)].shape

(34303, 6)

In [ ]:
high_depth_onsr= high_depth[high_depth['Address'].str.contains('onsr', case=False)]

In [ ]:
high_depth_onsr_html=high_depth_onsr[high_depth_onsr['Content Type'].str.contains('html', case=False)]
high_depth_onsr_html.shape

(12664, 6)

In [ ]:
df_onsr=df[df['Address'].str.contains('onsr', case=False)]
df_onsr.shape

(37345, 43)

In [ ]:
df_onsr_html=df_onsr[df_onsr['Content Type'].str.contains('html', case=False)]
df_onsr_html.shape

(13815, 43)

There are 12664 rows with Crawl Depth >1000
While checking the addresses of these rows I found out that all of them are from one website (onsr: https://onsr.nat.tn/onsr/...) and their content type is html.
While clicking on some of these links I didn't find any content they are all empty, I only found the principal menu and the side bar
--> All links related to onsr must be removed


In [ ]:
df = df[~df['Address'].str.contains('onsr', case=False)]

In [ ]:
df[df['Address'].str.contains('onsr', case=False)].shape

(0, 43)

In [ ]:
df.shape

(1095615, 43)

Now let's go back to Crawl Depth

## Back to Crawl Depth

Now the dataset doesen't contain in Crawl Depth > 1000, let's check the higher value it contains

In [ ]:
df['Crawl Depth'].max()

261.0

In [ ]:
df[df['Crawl Depth']==261]

Address Content Type  \
314184  http://www.utm.rnu.tn/visirech/Fr/utm/annuaire...         html   
314210  http://www.utm.rnu.tn/visirech/Fr/utm/annuaire...         html   

        Status Code Status Indexability Indexability Status   Title 1  \
314184          200     OK    Indexable                 NaN  VISIRECH   
314210          200     OK    Indexable                 NaN  VISIRECH   

        Title 1 Length  Title 1 Pixel Width  \
314184               8                   85   
314210               8                   85   

                                       Meta Description 1  ...  Redirect Type  \
314184  Plateforme de la recherche scientifique de l'U...  ...            NaN   
314210  Plateforme de la recherche scientifique de l'U...  ...            NaN   

        Cookies PDF Subject  PDF Author PDF Created  PDF Modified  \
314184        1         NaN         NaN         NaN           NaN   
314210        1         NaN         NaN         NaN           NaN   

       PDF Page Count  HTTP Version  \
314184              0           1.1   
314210              0           1.1   

                                      URL Encoded Address      Crawl Timestamp  
314184  http://www.utm.rnu.tn/visirech/Fr/utm/annuaire...  2023-07-28 18:32:51  
314210  http://www.utm.rnu.tn/visirech/Fr/utm/annuaire...  2023-07-28 18:32:51  

[2 rows x 43 columns]

In [ ]:
df['Crawl Depth'].min()

0.0

In [ ]:
df[df['Crawl Depth']==0].Address.head()

29275           http://mastere.utm.rnu.tn/
30792             https://www.apia.com.tn/
41641         https://www.social.gov.tn/fr
63118            https://www.oc.com.tn/fr/
74145    http://www.sport.tn/index.php/ar/
Name: Address, dtype: object

I don't know if we should remove rows with Crawl Depth = 0 because they lead to the homepage where it dosent necessarily contain relevant infos such as procedures, articles...

# Status column

In [ ]:
df['Status'].value_counts()

OK                                  1095089
Connexion refusée                       221
200                                     197
Expiration du délai de connexion        104
Name: Status, dtype: int64

After all the filtering we did above, now we only have rows with status OK and 200 so we don't need it for other filterings.

# Indexability and Indexability Status columns

In [ ]:
df['Indexability'].value_counts()

Indexable        973280
Non-Indexable    122335
Name: Indexability, dtype: int64

In [ ]:
df[['Indexability','Indexability Status']].value_counts()

Indexability   Indexability Status   
Non-Indexable  Canonicalised             116117
               noindex                     5981
               Canonicalised,noindex        144
               Max Page Size Exceeded        51
               Redirected                    42
dtype: int64

We notice that Indexability Status is filled only when Indexability = Non-Indexable so we're going to focus on that column.
After some research we found out that The canonical version is considered the original or primary page, and it is the one that search engines are more likely to show in search results.

In [ ]:
df[ df['Indexability Status']=='noindex'].head()

Address Content Type  \
31   http://www.sport.tn/index.php/ar/1/1140-الاطلا...         html   
111  http://www.health.gov.tn/fr/toutes-les-actuali...         html   
361  http://www.sport.tn/index.php/ar/component/con...         html   
762  http://www.sport.tn/index.php/ar/component/con...         html   
796  http://www.sport.tn/index.php/ar/أخبار-رئيسية/...         html   

     Status Code Status   Indexability Indexability Status  \
31           200     OK  Non-Indexable             noindex   
111          200     OK  Non-Indexable             noindex   
361          200     OK  Non-Indexable             noindex   
762          200     OK  Non-Indexable             noindex   
796          200     OK  Non-Indexable             noindex   

                                               Title 1  Title 1 Length  \
31   الاطلاع على ظروف المعهد العالي للرياضة والتربي...              63   
111  Concours sur épreuves pour la nomination au gr...             150   
361                              وزارة الشباب والرياضة              21   
762                              وزارة الشباب والرياضة              21   
796  الاطلاع على واقع الفضاءات الرياضية بالمدارس ال...              66   

     Title 1 Pixel Width Meta Description 1  ...  Redirect Type  Cookies  \
31                   522                NaN  ...            NaN        2   
111                 1357                NaN  ...            NaN        2   
361                  181                NaN  ...            NaN        2   
762                  181                NaN  ...            NaN        2   
796                  562                NaN  ...            NaN        2   

    PDF Subject  PDF Author PDF Created  PDF Modified PDF Page Count  \
31          NaN         NaN         NaN           NaN              0   
111         NaN         NaN         NaN           NaN              0   
361         NaN         NaN         NaN           NaN              0   
762         NaN         NaN         NaN           NaN              0   
796         NaN         NaN         NaN           NaN              0   

     HTTP Version                                URL Encoded Address  \
31            1.1  http://www.sport.tn/index.php/ar/1/1140-%D8%A7...   
111           1.1  http://www.health.gov.tn/fr/toutes-les-actuali...   
361           1.1  http://www.sport.tn/index.php/ar/component/con...   
762           1.1  http://www.sport.tn/index.php/ar/component/con...   
796           1.1  http://www.sport.tn/index.php/ar/%D8%A3%D8%AE%...   

         Crawl Timestamp  
31   2023-07-27 18:51:03  
111  2023-07-25 21:14:35  
361  2023-07-23 19:15:39  
762  2023-07-23 19:16:09  
796  2023-07-28 17:22:50  

[5 rows x 43 columns]

When checking data with Indexability Status=noindex we notice that most of it have noindex,nofollow in Meta Robots 1 column, most of it contain data from the health government website so we must check these two patterns.

## Meta Robots 1 column

In [ ]:
df.columns

Index(['Address', 'Content Type', 'Status Code', 'Status', 'Indexability',
       'Indexability Status', 'Title 1', 'Title 1 Length',
       'Title 1 Pixel Width', 'Meta Description 1',
       'Meta Description 1 Length', 'Meta Description 1 Pixel Width',
       'Meta Keywords 1', 'Meta Keywords 1 Length', 'H1-1', 'H1-1 Length',
       'H2-1', 'H2-1 Length', 'Meta Robots 1', 'X-Robots-Tag 1',
       'Meta Refresh 1', 'Size (bytes)', 'Word Count', 'Sentence Count',
       'Average Words Per Sentence', 'Flesch Reading Ease Score',
       'Readability', 'Text Ratio', 'Crawl Depth', 'Hash', 'Response Time',
       'Last Modified', 'Redirect URL', 'Redirect Type', 'Cookies',
       'PDF Subject', 'PDF Author', 'PDF Created', 'PDF Modified',
       'PDF Page Count', 'HTTP Version', 'URL Encoded Address',
       'Crawl Timestamp'],
      dtype='object')

In [ ]:
df['Meta Robots 1'].value_counts()

index,follow                                                                    77497
max-image-preview:large                                                          7283
noindex, nofollow                                                                5540
all                                                                              2316
index, follow                                                                    1096
index, follow, max-image-preview:large, max-snippet:-1, max-video-preview:-1      526
noodp,noydir                                                                      511
index, follow, max-snippet:-1, max-video-preview:-1, max-image-preview:large      258
noindex,follow                                                                    240
noindex,noodp,noydir                                                              126
follow, index, max-snippet:-1, max-video-preview:-1, max-image-preview:large       26
follow, noindex                                       

In [ ]:
df[df['Meta Robots 1'] == 'noydir']

Address Content Type  \
1108519  https://www.amideast.org/yemen/find-a-scholars...         html   

         Status Code Status Indexability Indexability Status  \
1108519          200     OK    Indexable                 NaN   

                                                  Title 1  Title 1 Length  \
1108519  [Undergraduate Scholarships] | [AMIDEAST][Yemen]              48   

         Title 1 Pixel Width Meta Description 1  ...  Redirect Type  Cookies  \
1108519                  456                NaN  ...            NaN        0   

        PDF Subject  PDF Author PDF Created  PDF Modified PDF Page Count  \
1108519         NaN         NaN         NaN           NaN              0   

         HTTP Version                                URL Encoded Address  \
1108519           1.1  https://www.amideast.org/yemen/find-a-scholars...   

             Crawl Timestamp  
1108519  2023-07-22 20:00:59  

[1 rows x 43 columns]

In [ ]:
df[df['Address'].str.contains('yemen', case=False)].shape

(119, 43)

*Amideast addresses*

In [ ]:
amideast_adr = df[df['Address'].str.contains('amideast', case=False)]
amideast_adr.shape

(2263, 43)

In [ ]:
amideast_adr[~amideast_adr['Address'].str.contains('tunisia')].shape

(2140, 43)

When observing the address that corresponds to the value 'noydir' of the Meta Robots 1 column, we can see that it's linked to amideast of yemen. After that I searched for every address that contains yemen in it, they are around 120.
Then for addresses that don't contain tunisia, I found that out of 2278 amideast link there are 2150 that aren't for tunisian students. These addresses should be removed.

In [ ]:
# Remove links of other countries
df = df[~(df['Address'].str.contains('amideast', case=False) & ~df['Address'].str.contains('tunisia', case=False))]

In [ ]:
df.shape

(1093477, 43)

*Back to Meta Robots 1*

In [ ]:
df['Meta Robots 1'].value_counts()

index,follow                                                                    77497
max-image-preview:large                                                          7283
noindex, nofollow                                                                5540
all                                                                              2316
index, follow                                                                    1096
index, follow, max-image-preview:large, max-snippet:-1, max-video-preview:-1      526
noodp,noydir                                                                      511
index, follow, max-snippet:-1, max-video-preview:-1, max-image-preview:large      258
noindex,follow                                                                    240
noindex,noodp,noydir                                                              126
follow, index, max-snippet:-1, max-video-preview:-1, max-image-preview:large       26
follow, noindex                                       

In [ ]:
df[df['Meta Robots 1'] == 'All']

Address Content Type  Status Code Status  \
248130  http://www.ismso.rnu.tn/         html          200     OK   

       Indexability Indexability Status  \
248130    Indexable                 NaN   

                                                  Title 1  Title 1 Length  \
248130  ...::: Institut supérieur de Musique de Sousse...              53   

        Title 1 Pixel Width Meta Description 1  ...  Redirect Type  Cookies  \
248130                  427                NaN  ...            NaN        0   

       PDF Subject  PDF Author PDF Created  PDF Modified PDF Page Count  \
248130         NaN         NaN         NaN           NaN              0   

        HTTP Version       URL Encoded Address      Crawl Timestamp  
248130           1.1  http://www.ismso.rnu.tn/  2023-07-25 19:26:37  

[1 rows x 43 columns]

In [ ]:
# Remove the above row
df = df[~(df['Meta Robots 1'] == 'All')]

In [ ]:
df['Meta Robots 1'].value_counts()

index,follow                                                                    77497
max-image-preview:large                                                          7283
noindex, nofollow                                                                5540
all                                                                              2316
index, follow                                                                    1096
index, follow, max-image-preview:large, max-snippet:-1, max-video-preview:-1      526
noodp,noydir                                                                      511
index, follow, max-snippet:-1, max-video-preview:-1, max-image-preview:large      258
noindex,follow                                                                    240
noindex,noodp,noydir                                                              126
follow, index, max-snippet:-1, max-video-preview:-1, max-image-preview:large       26
follow, noindex                                       

In [ ]:
replacement_dict = {
    'noindex, nofollow': 'noindex,nofollow',
    'NOINDEX, NOFOLLOW': 'noindex,nofollow',
    'noindex, follow': 'noindex,follow',
    'follow, noindex': 'noindex,follow',
    'noindex,noodp,noydir' : 'noindex',
    'index, follow, max-snippet:-1, max-video-preview:-1, max-image-preview:large': 'index,follow',
    'index, follow, max-image-preview:large, max-snippet:-1, max-video-preview:-1': 'index,follow',
    'index, follow': 'index,follow',
    'all':'index,follow'
    }
df['Meta Robots 1'] = df['Meta Robots 1'].replace(replacement_dict)

In [ ]:
df['Meta Robots 1'].value_counts()

index,follow                                                                    81693
max-image-preview:large                                                          7283
noindex,nofollow                                                                 5540
noodp,noydir                                                                      511
noindex,follow                                                                    252
noindex                                                                           131
follow, index, max-snippet:-1, max-video-preview:-1, max-image-preview:large       26
index                                                                               7
noindex,noarchive                                                                   1
Name: Meta Robots 1, dtype: int64

In [ ]:
df[df['Meta Robots 1'] == 'noindex,noarchive']

Address Content Type  Status Code Status  \
345062  https://www.ugtt.org.tn/pinglink/         html          200     OK   

         Indexability Indexability Status                             Title 1  \
345062  Non-Indexable             noindex  الاتحاد العام التونسي للشغل › دخول   

        Title 1 Length  Title 1 Pixel Width Meta Description 1  ...  \
345062              34                  283                NaN  ...   

        Redirect Type  Cookies PDF Subject  PDF Author PDF Created  \
345062            NaN        1         NaN         NaN         NaN   

        PDF Modified PDF Page Count  HTTP Version  \
345062           NaN              0           1.1   

                      URL Encoded Address      Crawl Timestamp  
345062  https://www.ugtt.org.tn/pinglink/  2023-07-22 18:02:04  

[1 rows x 43 columns]

In [ ]:
df = df[~(df['Meta Robots 1'] == 'noindex,noarchive')]

In [ ]:
df['Meta Robots 1'].value_counts()

index,follow                                                                    81693
max-image-preview:large                                                          7283
noindex,nofollow                                                                 5540
noodp,noydir                                                                      511
noindex,follow                                                                    252
noindex                                                                           131
follow, index, max-snippet:-1, max-video-preview:-1, max-image-preview:large       26
index                                                                               7
Name: Meta Robots 1, dtype: int64

In [ ]:
df[(df['Meta Robots 1'] == 'follow, index, max-snippet:-1, max-video-preview:-1, max-image-preview:large')].Address

490        http://www.utunis.rnu.tn/Conférance-Débat/Salo...
11827      http://www.utunis.rnu.tn/Conférance-Débat/Hass...
12389      http://www.utunis.rnu.tn/Conférance-Débat/Rim%...
17541      http://www.utunis.rnu.tn/Conférance-Débat/Ali%...
31781      http://www.utunis.rnu.tn/Conférance-Débat/Habi...
85440      http://www.utunis.rnu.tn/Conférance-Débat/Hich...
179827     http://www.utunis.rnu.tn/Conférance-Débat/Monc...
189819     http://www.utunis.rnu.tn/Conférance-Débat/Hass...
255090     http://www.utunis.rnu.tn/Conférance-Débat/Moha...
257258     http://www.utunis.rnu.tn/Conférance-Débat/Noub...
387538     http://www.utunis.rnu.tn/Conférance-Débat/Mejd...
590162     http://www.utunis.rnu.tn/Conférance-Débat/Nafâ...
622409     http://www.utunis.rnu.tn/Conférance-Débat/Imen...
632494     http://www.utunis.rnu.tn/Conférance-Débat/Sarr...
717345     http://www.utunis.rnu.tn/Conférance-Débat/Fara...
743871     http://www.utunis.rnu.tn/Conférance-Débat/Mour...
759092     http://www.ut

In [ ]:
df[df['Address'].str.contains('utunis',case='False')].shape

(317, 43)

In [ ]:
df[df['Address'].str.contains('utunis',case='False') & df['Address'].str.contains('Débat',case='False')].shape

(28, 43)

In [ ]:
# Remove links that contain 'utunis' and 'débat'
df = df[~(df['Address'].str.contains('utunis',case='False') & df['Address'].str.contains('Débat',case='False'))]

In [ ]:
df[df['Address'].str.contains('utunis',case='False')].shape

(289, 43)

In [ ]:
df['Meta Robots 1'].value_counts()

index,follow               81693
max-image-preview:large     7283
noindex,nofollow            5540
noodp,noydir                 511
noindex,follow               252
noindex                      131
index                          7
Name: Meta Robots 1, dtype: int64

In [ ]:
# Remove rows with Meta Robots 1 = noindex,nofollow
df = df[df['Meta Robots 1'] != 'noindex,nofollow']

In [ ]:
df.shape

(1087907, 43)

## Health website

In [ ]:
health_data = df[df['Address'].str.contains('health', case=False)]
health_data.shape

(1758, 43)

In [ ]:
health_data.groupby('Meta Robots 1', dropna=False, as_index=False)['Address'].count()

Meta Robots 1  Address
0  index,follow       13
1           NaN     1745

In [ ]:
health_data[health_data['Meta Robots 1']=='index,follow'].Address

28821      http://www.pasteur.tn/index.php?option=com_con...
238036     http://www.pasteur.tn/index.php?option=com_con...
276145     http://www.pasteur.tn/index.php?option=com_con...
555847     http://www.pasteur.tn/index.php?option=com_con...
692218     http://www.pasteur.tn/index.php?option=com_con...
769076     http://www.pasteur.tn/index.php?option=com_con...
775733     http://www.pasteur.tn/index.php?option=com_con...
979298        https://istmt.rnu.tn/hacking-health-sfax-2018/
1010764    http://www.pasteur.tn/index.php?option=com_con...
1057837    https://istmt.rnu.tn/en/tunisia-ehealth-challe...
1190675    http://www.pasteur.tn/index.php?option=com_con...
1420335    https://istmt.rnu.tn/en/hacking-health-sfax-20...
1435487    https://istmt.rnu.tn/tunisia-ehealth-challenge...
Name: Address, dtype: object

In [ ]:
# I noticed addresses that have pasteur and option are all deprecated let's check them more
pasteur = df[df['Address'].str.contains('pasteur',case=False) & df['Address'].str.contains('option',case=False)]
pasteur.shape

(824, 43)

In [ ]:
pasteur.Address.sample(10)

219362     http://www.pasteur.tn/index.php?option=com_con...
1215013    http://www.pasteur.tn/index.php?option=com_doc...
239992     http://www.pasteur.tn/index.php?option=com_con...
802707     http://www.pasteur.tn/index.php?option=com_doc...
1095236    http://www.pasteur.tn/index.php?option=com_con...
1219092    http://www.pasteur.tn/index.php?option=com_doc...
803619     http://www.pasteur.tn/index.php?option=com_con...
779966     http://www.pasteur.tn/index.php?option=com_con...
254556     http://www.pasteur.tn/index.php?option=com_con...
1404765    http://www.pasteur.tn/index.php?option=com_doc...
Name: Address, dtype: object

In [ ]:
health_data = health_data[~(health_data['Address'].str.contains('pasteur', case=False) & health_data['Address'].str.contains('option', case=False))]
df = df[~(df['Address'].str.contains('pasteur', case=False) & df['Address'].str.contains('option', case=False))]
df.shape

(1087083, 43)

In [ ]:
health_data[health_data['Meta Robots 1']=='index,follow'].Address

979298        https://istmt.rnu.tn/hacking-health-sfax-2018/
1057837    https://istmt.rnu.tn/en/tunisia-ehealth-challe...
1420335    https://istmt.rnu.tn/en/hacking-health-sfax-20...
1435487    https://istmt.rnu.tn/tunisia-ehealth-challenge...
Name: Address, dtype: object

In [ ]:
# The links above are irrelevant
df = df[~((df['Meta Robots 1']=='index,follow') & df['Address'].str.contains('health'))]
health_data = health_data[health_data['Meta Robots 1']!='index,follow']
df.shape

(1087079, 43)

In [ ]:
health_data.groupby('Meta Robots 1', dropna=False, as_index=False)['Address'].count()

Meta Robots 1  Address
0            NaN     1745

In [ ]:
health_data.Address.sample(10)

44185      http://www.health.gov.tn/fr/toutes-les-actuali...
771796     http://www.health.gov.tn/fr/articles-archivés/...
1210895    http://www.health.gov.tn/fr/sante-en-tunisie/s...
733567     http://www.health.gov.tn/fr/articles-archivés/...
550934     http://www.health.gov.tn/fr/toutes-les-actuali...
872054     http://www.health.gov.tn/fr/articles-archivés/...
390020     http://www.health.gov.tn/fr/articles-archivés/...
50727      http://www.health.gov.tn/fr/prestations/cahier...
376043     http://www.health.gov.tn/fr/articles-archivés/...
1306419    http://www.health.gov.tn/fr/articles-archivés/...
Name: Address, dtype: object


After some research we found out that many links related to health data are deprecated and should be removed.

In [ ]:
df = df[~df['Address'].str.contains('health', case=False)]

In [ ]:
df.shape

(1085334, 43)

## Back to Indexability and Indexability Status columns

In [ ]:
df[['Indexability','Indexability Status']].value_counts()

Indexability   Indexability Status   
Non-Indexable  Canonicalised             116047
               noindex                      440
               Canonicalised,noindex        144
               Max Page Size Exceeded        50
               Redirected                    42
dtype: int64

In [ ]:
df[(df['Indexability'] == 'Non-Indexable') & (df['Indexability Status'] == 'Redirected')].Address

61156                    https://www.atct.tn/rh/fr/cooperant
86386      http://rel.uvt.rnu.tn/pluginfile.php/3339/mod_...
90322      http://rel.uvt.rnu.tn/pluginfile.php/3323/mod_...
187319      https://www.atct.tn/rh/fr/consultant/inscription
202525                  https://www.atct.tn/rh/en/consultant
298290                   https://www.atct.tn/rh/en/employeur
324304                             http://www.isetkr.rnu.tn/
326658                             https://www.atct.tn/rh/ar
326743                             https://www.atct.tn/rh/fr
326783                             https://www.atct.tn/rh/en
343057        https://www.atct.tn/rh/fr/candidat/inscription
406290     http://rel.uvt.rnu.tn/pluginfile.php/3317/mod_...
412354     http://rel.uvt.rnu.tn/pluginfile.php/3311/mod_...
452893                    https://www.atct.tn/rh/fr/candidat
464573                   https://www.atct.tn/rh/ar/cooperant
513085                           http://www.sphb.defense.tn/
527516                  

In [ ]:
df[df['Address'].str.contains('pluginfile')].Address.sample(10)

1320449    http://rel.uvt.rnu.tn/pluginfile.php/19372/mod...
1374932    http://rel.uvt.rnu.tn/pluginfile.php/19372/mod...
805762     http://rel.uvt.rnu.tn/pluginfile.php/3344/mod_...
400699     http://rel.uvt.rnu.tn/pluginfile.php/3317/mod_...
1460892    http://rel.uvt.rnu.tn/pluginfile.php/3300/mod_...
1268977    http://rel.uvt.rnu.tn/pluginfile.php/3333/mod_...
1401504    http://rel.uvt.rnu.tn/pluginfile.php/3320/mod_...
1409072    http://rel.uvt.rnu.tn/pluginfile.php/3333/mod_...
417247     http://rel.uvt.rnu.tn/pluginfile.php/3320/mod_...
251960     http://rel.uvt.rnu.tn/pluginfile.php/3306/mod_...
Name: Address, dtype: object

In [ ]:
# The links above are deprecated
df = df[~df['Address'].str.contains('pluginfile', case=False)]

In [ ]:
# These links are deprecated
df = df[~((df['Address'].str.contains('mdici')) & (df['Address'].str.contains('المستجدات')))]

In [ ]:
df.shape

(1084705, 43)

# Check missing data

In [ ]:
df.shape

(1084705, 43)

In [ ]:
df.isnull().sum().sort_values(ascending=False)

Redirect Type                     1084705
Redirect URL                      1084705
Meta Refresh 1                    1084676
X-Robots-Tag 1                    1084507
PDF Subject                       1084332
PDF Author                        1071208
PDF Modified                      1064949
PDF Created                       1063456
Meta Robots 1                      995556
Indexability Status                968038
Last Modified                      935452
H1-1                               839907
H2-1                               671393
Meta Keywords 1                    327505
Meta Description 1                 257921
Title 1                             46920
Sentence Count                      42243
Average Words Per Sentence          42243
Flesch Reading Ease Score           42243
Readability                         42243
Hash                                39598
Crawl Depth                          4180
Status                                  4
Cookies                           

These columns: 'Redirect URL', 'Redirect Type', 'Meta Refresh 1', 'X-Robots-Tag 1' must be dropped because they contain many missing values.

In [ ]:
columns_to_drop = ['Redirect URL', 'Redirect Type', 'Meta Refresh 1', 'X-Robots-Tag 1', 'Crawl Timestamp', 'URL Encoded Address', 'HTTP Version', 'Cookies']
df.drop(columns_to_drop, axis=1, inplace=True)

In [ ]:
df.isnull().sum().sort_values(ascending=False)

PDF Subject                       1084332
PDF Author                        1071208
PDF Modified                      1064949
PDF Created                       1063456
Meta Robots 1                      995556
Indexability Status                968038
Last Modified                      935452
H1-1                               839907
H2-1                               671393
Meta Keywords 1                    327505
Meta Description 1                 257921
Title 1                             46920
Average Words Per Sentence          42243
Readability                         42243
Flesch Reading Ease Score           42243
Sentence Count                      42243
Hash                                39598
Crawl Depth                          4180
Status                                  4
Address                                 0
Text Ratio                              0
Response Time                           0
H2-1 Length                             0
Word Count                        

In [ ]:
df.loc[df['H1-1'].isna(), 'H1-1 Length'] = pd.NA

In [ ]:
df.loc[df['H2-1'].isna(), 'H2-1 Length'] = pd.NA

In [ ]:
df.loc[df['Meta Keywords 1'].isna(), 'Meta Keywords 1 Length'] = pd.NA

In [ ]:
df.loc[df['Meta Description 1'].isna(), ['Meta Description 1 Length', 'Meta Description 1 Pixel Width']] = pd.NA

In [ ]:
df.loc[df['Title 1'].isna(), ['Title 1 Pixel Width','Title 1 Length']] = pd.NA

In [ ]:
df.isnull().sum().sort_values()

Address                                 0
Response Time                           0
Text Ratio                              0
Word Count                              0
Size (bytes)                            0
PDF Page Count                          0
Content Type                            0
Status Code                             0
Indexability                            0
Status                                  4
Crawl Depth                          4180
Hash                                39598
Readability                         42243
Flesch Reading Ease Score           42243
Average Words Per Sentence          42243
Sentence Count                      42243
Title 1                             46920
Title 1 Length                      46920
Title 1 Pixel Width                 46920
Meta Description 1 Length          257921
Meta Description 1                 257921
Meta Description 1 Pixel Width     257921
Meta Keywords 1 Length             327505
Meta Keywords 1                   

# Notes

- Other unecessary columns: 'Crawl Timestamp', 'URL Encoded Address', 'HTTP Version', 'Cookies'
- Notes: While observing the different rows of the columns Address and URL Encoded Address they leaded to the same link.
- Drop rows with content type != html and pdf
- Drop rows with hash duplicated (keep first)
- After that check Indexability and Index. Status columns
- All columns with pixel width may be useless.
- Title 1 Length = 638 (max) is irrelevant
- df['Title 1 Pixel Width']==0 correspond to the missing values in column Title 1, same for Title 1 Length and then check the rest of the rows
- Addresses with Meta Description 1 Length = 0 are NaN and = 3 are irrelevant.
- Addresses with Meta Description 1 Pixel Width>4000 are dysfunctionnal.
- I'm not sure wether the column meta keywords is useful or not, some contain logical keywords and some don't (example: (), l, etc)
- NaN meta keywords should have NaN meta keywords length (replace 0 with NaN)
- Most of the addresses with H2-1 Length > 300 contain long articles without any document or legal procedures.
- We can filter based on titles and descriptions using words like document, imprimés ...
- Rows with H2-1 Length = 829 to drop.
- Addresses that have H2-1 Length = 1 are dysfunctional.
- There are missing values in H1-1 and H2-1 but there aren't any in their lengths (replace 0 with NaN).
- Size Bytes (0 may mean empty content)
- Rows with Word Count = 0 must be removed (but we have first to check its content type)
- Columns with small number of Sentence Count must be removed (check content type) + Average Words Per Sentence  
- Check for indirect duplicates.


- Size (bytes) column: there are more than 19000 rows with size = 0, after trying the links I found out that links with content type=html are functional whereas links with type pdf are dysfunctionnal.
- There are many addresses with status code 200 and status ok but then they show 400 not found





# Dataset Division

In [ ]:
df['Content Type'].value_counts()

html    1045429
pdf       39276
Name: Content Type, dtype: int64

In [ ]:
pdf = df[df['Content Type'] == 'pdf']
pdf.shape

(39276, 35)

In [ ]:
html = df[df['Content Type'] == 'html']
html.shape

(1045429, 35)

Now let's focus on each dataset seperately

## PDF dataset

In [ ]:
pdf.describe()

Status Code  Title 1 Length  Title 1 Pixel Width  \
count      39276.0     8643.000000          8643.000000   
mean         200.0       34.143584           320.824251   
std            0.0       28.322176           286.503534   
min          200.0        1.000000             0.000000   
25%          200.0       17.000000           157.000000   
50%          200.0       26.000000           259.000000   
75%          200.0       43.000000           395.000000   
max          200.0      343.000000          3250.000000   

       Meta Description 1 Length  Meta Description 1 Pixel Width  \
count                        0.0                             0.0   
mean                         NaN                             NaN   
std                          NaN                             NaN   
min                          NaN                             NaN   
25%                          NaN                             NaN   
50%                          NaN                             NaN   
75%                          NaN                             NaN   
max                          NaN                             NaN   

       Meta Keywords 1 Length  H1-1 Length  H2-1 Length  Size (bytes)  \
count              127.000000          0.0          0.0  3.927600e+04   
mean                13.267717          NaN          NaN  5.401944e+05   
std                 14.651750          NaN          NaN  4.096860e+06   
min                  1.000000          NaN          NaN  0.000000e+00   
25%                  4.000000          NaN          NaN  0.000000e+00   
50%                 11.000000          NaN          NaN  3.129600e+04   
75%                 15.000000          NaN          NaN  2.532802e+05   
max                106.000000          NaN          NaN  3.385301e+08   

       Word Count  Sentence Count  Average Words Per Sentence  \
count     39276.0             0.0                         0.0   
mean          0.0             NaN                         NaN   
std           0.0             NaN                         NaN   
min           0.0             NaN                         NaN   
25%           0.0             NaN                         NaN   
50%           0.0             NaN                         NaN   
75%           0.0             NaN                         NaN   
max           0.0             NaN                         NaN   

       Flesch Reading Ease Score  Text Ratio   Crawl Depth  Response Time  \
count                        0.0     39276.0  38470.000000   39276.000000   
mean                         NaN         0.0     10.705641       0.168429   
std                          NaN         0.0     18.413720       0.456006   
min                          NaN         0.0      1.000000       0.008000   
25%                          NaN         0.0      3.000000       0.052000   
50%                          NaN         0.0      4.000000       0.111000   
75%                          NaN         0.0      9.000000       0.183000   
max                          NaN         0.0    183.000000      20.234000   

       PDF Page Count  
count    39276.000000  
mean         4.636852  
std         19.214837  
min          0.000000  
25%          0.000000  
50%          1.000000  
75%          3.000000  
max        621.000000

In [ ]:
pdf.isnull().sum()

Address                               0
Content Type                          0
Status Code                           0
Status                                0
Indexability                          0
Indexability Status               39038
Title 1                           30633
Title 1 Length                    30633
Title 1 Pixel Width               30633
Meta Description 1                39276
Meta Description 1 Length         39276
Meta Description 1 Pixel Width    39276
Meta Keywords 1                   39149
Meta Keywords 1 Length            39149
H1-1                              39276
H1-1 Length                       39276
H2-1                              39276
H2-1 Length                       39276
Meta Robots 1                     39276
Size (bytes)                          0
Word Count                            0
Sentence Count                    39276
Average Words Per Sentence        39276
Flesch Reading Ease Score         39276
Readability                       39276


In [ ]:
# Remove columns with more than 39000 missing values
pdf.dropna(thresh=(len(pdf)-39000), axis=1, inplace=True)

<ipython-input-126-8f2e06c41c06>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  pdf.dropna(thresh=(len(pdf)-39000), axis=1, inplace=True)


In [ ]:
pdf.isna().sum()

Address                    0
Content Type               0
Status Code                0
Status                     0
Indexability               0
Title 1                30633
Title 1 Length         30633
Title 1 Pixel Width    30633
Size (bytes)               0
Word Count                 0
Text Ratio                 0
Crawl Depth              806
Response Time              0
Last Modified           9427
PDF Subject            38903
PDF Author             25779
PDF Created            18027
PDF Modified           19520
PDF Page Count             0
dtype: int64

In [ ]:
# Remove unneeded columns
unneeded = ['Title 1', 'Title 1 Length', 'Title 1 Pixel Width', 'Word Count', 'Text Ratio', 'Response Time', 'Content Type', 'Status Code', 'Status']
pdf.drop(unneeded, axis=1, inplace=True)

<ipython-input-128-185c5ff9fff6>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  pdf.drop(unneeded, axis=1, inplace=True)


In [ ]:
pdf.columns

Index(['Address', 'Indexability', 'Size (bytes)', 'Crawl Depth',
       'Last Modified', 'PDF Subject', 'PDF Author', 'PDF Created',
       'PDF Modified', 'PDF Page Count'],
      dtype='object')

## HTML dataset

In [ ]:
html.isna().sum()

Address                                 0
Content Type                            0
Status Code                             0
Status                                  4
Indexability                            0
Indexability Status                929000
Title 1                             16287
Title 1 Length                      16287
Title 1 Pixel Width                 16287
Meta Description 1                 218645
Meta Description 1 Length          218645
Meta Description 1 Pixel Width     218645
Meta Keywords 1                    288356
Meta Keywords 1 Length             288356
H1-1                               800631
H1-1 Length                        800631
H2-1                               632117
H2-1 Length                        632117
Meta Robots 1                      956280
Size (bytes)                            0
Word Count                              0
Sentence Count                       2967
Average Words Per Sentence           2967
Flesch Reading Ease Score         

In [ ]:
# Remove columns with 100% NaN values
html.dropna(how='all', axis=1, inplace=True)

<ipython-input-131-e236ad3206ca>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  html.dropna(how='all', axis=1, inplace=True)


In [ ]:
html = html[~html['Status'].isna()]

In [ ]:
html.isna().sum()

Address                                0
Content Type                           0
Status Code                            0
Status                                 0
Indexability                           0
Indexability Status               928996
Title 1                            16286
Title 1 Length                     16286
Title 1 Pixel Width                16286
Meta Description 1                218641
Meta Description 1 Length         218641
Meta Description 1 Pixel Width    218641
Meta Keywords 1                   288352
Meta Keywords 1 Length            288352
H1-1                              800628
H1-1 Length                       800628
H2-1                              632113
H2-1 Length                       632113
Meta Robots 1                     956276
Size (bytes)                           0
Word Count                             0
Sentence Count                      2965
Average Words Per Sentence          2965
Flesch Reading Ease Score           2965
Readability     

In [ ]:
# Drop unnecessary columns
columns_to_drop = ['Content Type', 'Status Code', 'Crawl Depth', 'Hash', 'PDF Page Count', 'Response Time']
html.drop(columns_to_drop, axis=1, inplace=True)

In [ ]:
html['Status'].value_counts(dropna=False)

OK                                  1045061
Connexion refusée                       221
Expiration du délai de connexion        101
200                                      42
Name: Status, dtype: int64

In [ ]:
html['Status'] = html['Status'].str.replace('200','OK')

In [ ]:
# Remove rows with Status = Connexion refusée
html = html[~(html['Status'] == 'Connexion refusée')]

In [ ]:
html['Status'].value_counts(dropna=False)

OK                                  1045103
Expiration du délai de connexion        101
Name: Status, dtype: int64

In [ ]:
html['Indexability'].value_counts(dropna=False)

Indexable        928775
Non-Indexable    116429
Name: Indexability, dtype: int64

In [ ]:
html[['Indexability','Indexability Status']].value_counts(dropna=False)

Indexability   Indexability Status  
Indexable      NaN                      928775
Non-Indexable  Canonicalised            116047
               noindex                     251
               Canonicalised,noindex       103
               Redirected                   28
dtype: int64

In [ ]:
html[['Indexability','Meta Robots 1']].value_counts(dropna=False)

Indexability   Meta Robots 1          
Indexable      NaN                        840648
Non-Indexable  NaN                        115407
Indexable      index,follow                80951
               max-image-preview:large      6755
Non-Indexable  max-image-preview:large       528
Indexable      noodp,noydir                  414
Non-Indexable  noindex,follow                252
               noodp,noydir                   97
               noindex                        89
               index,follow                   56
Indexable      index                           7
dtype: int64

In [ ]:
html = html[~html['Address'].str.contains('facebook', case=False)]
html = html[~html['Address'].str.contains('twitter', case=False)]

In [ ]:
html.columns

Index(['Address', 'Status', 'Indexability', 'Indexability Status', 'Title 1',
       'Title 1 Length', 'Title 1 Pixel Width', 'Meta Description 1',
       'Meta Description 1 Length', 'Meta Description 1 Pixel Width',
       'Meta Keywords 1', 'Meta Keywords 1 Length', 'H1-1', 'H1-1 Length',
       'H2-1', 'H2-1 Length', 'Meta Robots 1', 'Size (bytes)', 'Word Count',
       'Sentence Count', 'Average Words Per Sentence',
       'Flesch Reading Ease Score', 'Readability', 'Text Ratio',
       'Last Modified'],
      dtype='object')

In [ ]:
html.shape

(1045122, 25)

In order to create two html datasets (High-Priority HTML and Low-Priority HTM), first of all we'll filter them based on this criteria:

1) High-Priority HTML have:
- 'Status = OK
- 'Indexability' = Indexable   
- 'Meta Robots 1' = index,follow

2) Low-Priority HTML contains the rest of the dataset

---

Then we'll determine other caracteristics to filter out the High-Priority HTML: This dataset will consist (somehow) of URLs that contain the most critical data.

Criteria (we have to create our formula for this):

- Word counts
- Sentence counts
- The ratio of H2 to H1 tags
- Originating URLs (maybe from Step 0 -data collection- ?)
- Other relevant metrics

We aim to apply strict criteria to ensure this tier's data quality. This dataset should not exceed 50,000 rows.


        

In [ ]:
html['Meta Robots 1'].value_counts()

index,follow               81005
max-image-preview:large     7283
noodp,noydir                 511
noindex,follow               252
noindex                       89
index                          7
Name: Meta Robots 1, dtype: int64

In [ ]:
html.columns

Index(['Address', 'Status', 'Indexability', 'Indexability Status', 'Title 1',
       'Title 1 Length', 'Title 1 Pixel Width', 'Meta Description 1',
       'Meta Description 1 Length', 'Meta Description 1 Pixel Width',
       'Meta Keywords 1', 'Meta Keywords 1 Length', 'H1-1', 'H1-1 Length',
       'H2-1', 'H2-1 Length', 'Meta Robots 1', 'Size (bytes)', 'Word Count',
       'Sentence Count', 'Average Words Per Sentence',
       'Flesch Reading Ease Score', 'Readability', 'Text Ratio',
       'Last Modified'],
      dtype='object')

## High-Priority HTML

In [ ]:
# Mark duplicated rows
html['Is_Duplicate'] = html.duplicated(subset=['Title 1', 'Meta Description 1', 'Meta Keywords 1', 'H1-1', 'H2-1'], keep=False)

# Count occurrences of each duplicated row
duplicate_counts = html['Is_Duplicate'].value_counts()

# Print duplicates count for each row
print(duplicate_counts)

True     978677
False     66445
Name: Is_Duplicate, dtype: int64


In [ ]:
html[(html['Is_Duplicate']==True) & (html['Address'].str.contains('interieur', case=False))].head()

Address Status Indexability  \
3  http://opendata.interieur.gov.tn/ar/datas/inde...     OK    Indexable   
4  http://opendata.interieur.gov.tn/fr/outils/ind...     OK    Indexable   
7  http://opendata.interieur.gov.tn/ar/catalog/re...     OK    Indexable   
8  http://opendata.interieur.gov.tn/ar/outils/ind...     OK    Indexable   
9  http://opendata.interieur.gov.tn/fr/outils/ind...     OK    Indexable   

  Indexability Status                                            Title 1  \
3                 NaN             موقع البيانات المفتوحة لوزارة الداخلية   
4                 NaN                                              قائمة   
7                 NaN  توزيع حوادث الطرقات وضحاياها خلال شهر أوت 2014...   
8                 NaN                                              قائمة   
9                 NaN                                              قائمة   

   Title 1 Length  Title 1 Pixel Width  \
3            38.0                326.0   
4             5.0                 45.0   
7            51.0                446.0   
8             5.0                 45.0   
9             5.0                 45.0   

                                  Meta Description 1  \
3  Le site officiel des données ouvertes du Minis...   
4  Le site officiel des données ouvertes du Minis...   
7  Le site officiel des données ouvertes du Minis...   
8  Le site officiel des données ouvertes du Minis...   
9  Le site officiel des données ouvertes du Minis...   

   Meta Description 1 Length  Meta Description 1 Pixel Width  ...  \
3                      120.0                           710.0  ...   
4                      120.0                           710.0  ...   
7                      120.0                           710.0  ...   
8                      120.0                           710.0  ...   
9                      120.0                           710.0  ...   

  Meta Robots 1  Size (bytes) Word Count  Sentence Count  \
3           NaN         55755        621           184.0   
4           NaN         51961        929           208.0   
7           NaN         43159        379           108.0   
8           NaN         50914        871           194.0   
9           NaN         51285        929           208.0   

  Average Words Per Sentence  Flesch Reading Ease Score  Readability  \
3                      3.375                     100.00    Very Easy   
4                      4.466                      59.42  Fairly Hard   
7                      3.509                     100.00    Very Easy   
8                      4.490                     100.00    Very Easy   
9                      4.466                      59.42  Fairly Hard   

   Text Ratio  Last Modified  Is_Duplicate  
3       8.492            NaT          True  
4      13.425            NaT          True  
7       7.319            NaT          True  
8      12.109            NaT          True  
9      13.602            NaT          True  

[5 rows x 26 columns]

In [ ]:
# To avoid comparing NaN values while removing duplicates I put them in a dataframe to see if they are relevant or not
nan_useful_links = html[html['Title 1'].isna() & html['Meta Description 1'].isna() & html['Meta Keywords 1'].isna() & html['H1-1'].isna() & html['H2-1'].isna()]

In [ ]:
nan_useful_links.sample(5)

Address Status  \
719955   https://www.orientini.com/go/8966concours.edunet.     OK   
1154484  https://www.orientini.com/go/2782safeonline.najah     OK   
1386000  https://www.orientini.com/go/2256russia.study_ar_     OK   
1482641  https://intranet.mehat.gov.tn/AccessInformatio...     OK   
701969   https://www.orientini.com/go/4904isigk.rnu.tn_fra     OK   

        Indexability Indexability Status Title 1  Title 1 Length  \
719955     Indexable                 NaN     NaN             NaN   
1154484    Indexable                 NaN     NaN             NaN   
1386000    Indexable                 NaN     NaN             NaN   
1482641    Indexable                 NaN     NaN             NaN   
701969     Indexable                 NaN     NaN             NaN   

         Title 1 Pixel Width Meta Description 1  Meta Description 1 Length  \
719955                   NaN                NaN                        NaN   
1154484                  NaN                NaN                        NaN   
1386000                  NaN                NaN                        NaN   
1482641                  NaN                NaN                        NaN   
701969                   NaN                NaN                        NaN   

         Meta Description 1 Pixel Width  ... Meta Robots 1  Size (bytes)  \
719955                              NaN  ...           NaN           172   
1154484                             NaN  ...           NaN           201   
1386000                             NaN  ...           NaN           178   
1482641                             NaN  ...           NaN        240750   
701969                              NaN  ...           NaN           151   

        Word Count  Sentence Count Average Words Per Sentence  \
719955           0             NaN                        NaN   
1154484          0             NaN                        NaN   
1386000          0             NaN                        NaN   
1482641       1159           627.0                      1.848   
701969           0             NaN                        NaN   

         Flesch Reading Ease Score Readability  Text Ratio  Last Modified  \
719955                         NaN         NaN       0.000            NaT   
1154484                        NaN         NaN       0.000            NaT   
1386000                        NaN         NaN       0.000            NaT   
1482641                      100.0   Very Easy       4.408            NaT   
701969                         NaN         NaN       0.000            NaT   

         Is_Duplicate  
719955           True  
1154484          True  
1386000          True  
1482641          True  
701969           True  

[5 rows x 26 columns]

In [ ]:
html_no_dup = html[~(html['Title 1'].isna() & html['Meta Description 1'].isna() & html['Meta Keywords 1'].isna() & html['H1-1'].isna() & html['H2-1'].isna())]

In [ ]:
html_no_dup.shape

(1042442, 26)

In [ ]:
html_no_dup.drop_duplicates(subset=['Title 1', 'Meta Description 1', 'Meta Keywords 1', 'H1-1', 'H2-1'],keep='first', inplace=True)
html_no_dup.shape

<ipython-input-153-992a8e548abf>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  html_no_dup.drop_duplicates(subset=['Title 1', 'Meta Description 1', 'Meta Keywords 1', 'H1-1', 'H2-1'],keep='first', inplace=True)


(115974, 26)

In [ ]:
html_no_dup[html_no_dup['Title 1'].isna() & html['Meta Description 1'].isna() & html['Meta Keywords 1'].isna() & html['H1-1'].isna() & html['H2-1'].isna() ].shape

<ipython-input-154-263c45330010>:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  html_no_dup[html_no_dup['Title 1'].isna() & html['Meta Description 1'].isna() & html['Meta Keywords 1'].isna() & html['H1-1'].isna() & html['H2-1'].isna() ].shape


(0, 26)

In [ ]:
del html_no_dup['Is_Duplicate']

In [ ]:
hp_html = html_no_dup.copy()

In [ ]:
hp_html.shape

(115974, 25)

In [ ]:
hp_html.columns

Index(['Address', 'Status', 'Indexability', 'Indexability Status', 'Title 1',
       'Title 1 Length', 'Title 1 Pixel Width', 'Meta Description 1',
       'Meta Description 1 Length', 'Meta Description 1 Pixel Width',
       'Meta Keywords 1', 'Meta Keywords 1 Length', 'H1-1', 'H1-1 Length',
       'H2-1', 'H2-1 Length', 'Meta Robots 1', 'Size (bytes)', 'Word Count',
       'Sentence Count', 'Average Words Per Sentence',
       'Flesch Reading Ease Score', 'Readability', 'Text Ratio',
       'Last Modified'],
      dtype='object')

In [ ]:
hp_html = hp_html[hp_html['Indexability'] == 'Indexable']

In [ ]:
hp_html.shape

(88158, 25)

In [ ]:
hp_html.isna().sum()

Address                               0
Status                                0
Indexability                          0
Indexability Status               88158
Title 1                             875
Title 1 Length                      875
Title 1 Pixel Width                 875
Meta Description 1                28145
Meta Description 1 Length         28145
Meta Description 1 Pixel Width    28145
Meta Keywords 1                   53572
Meta Keywords 1 Length            53572
H1-1                              25955
H1-1 Length                       25955
H2-1                              11135
H2-1 Length                       11135
Meta Robots 1                     77904
Size (bytes)                          0
Word Count                            0
Sentence Count                       29
Average Words Per Sentence           29
Flesch Reading Ease Score            29
Readability                          29
Text Ratio                            0
Last Modified                     63430


In [ ]:
hp_html[hp_html['Sentence Count']<=20].sample(10)

Address Status Indexability  \
203668  http://www.promotion.edunet.tn/2014/integsurgp...     OK    Indexable   
280599  http://www.onpc.nat.tn/index.php/ar/component/...     OK    Indexable   
349421  https://ent.uvt.rnu.tn/uvt/course/info.php?id=...     OK    Indexable   
24971   https://www.uvt.rnu.tn/formations/formations-i...     OK    Indexable   
342780  http://www.onpc.nat.tn/index.php/ar/component/...     OK    Indexable   
191977  http://www.onpc.nat.tn/index.php/ar/component/...     OK    Indexable   
976334  https://ent.uvt.rnu.tn/uvt/course/info.php?id=...     OK    Indexable   
672691                  https://www.innorpi.tn/en/node/31     OK    Indexable   
34084   http://www.onpc.nat.tn/index.php/ar/component/...     OK    Indexable   
381147  http://www.hmp.defense.tn/index.php/fr/site-ma...     OK    Indexable   

       Indexability Status                                            Title 1  \
203668                 NaN  إدماج القيمين العامين صنف (أ2) الحاملين للشهاد...   
280599                 NaN  قابس / مارث 02-02-2021 الساعة 08:09 حادث مرور ...   
349421                 NaN                             ملخص لـ  Ayouni wahida   
24971                  NaN                          UVT - Licences appliquées   
342780                 NaN  تونس/حلق الوادى الساعة 22:15 يوم 11/12/2022 مع...   
191977                 NaN      الكاف / نبر 16/10/2021 الساعة 19:37 حادث مرور   
976334                 NaN                   Résumé de Les compétences de vie   
672691                 NaN  Consultation N° 01/2014 Location d’un local à ...   
34084                  NaN  السواسي/المهدية يوم14/04/2023 الساعة 10:03  حر...   
381147                 NaN                                         Tacfarinas   

        Title 1 Length  Title 1 Pixel Width  \
203668           125.0               1041.0   
280599            50.0                431.0   
349421            22.0                193.0   
24971             25.0                235.0   
342780            54.0                475.0   
191977            45.0                394.0   
976334            32.0                321.0   
672691            80.0                736.0   
34084             54.0                474.0   
381147            10.0                 95.0   

                                       Meta Description 1  \
203668                                                NaN   
280599  جد حادث مرور بمدخل مدينة كتانة من معتمدية مارث...   
349421                                                NaN   
24971                       Université Virtuelle de Tunis   
342780  تولت وحدات النجدة و الإنقاذ التابعة للإدارة ال...   
191977  جد حادث مرور على مستوى منطقة بئر بن شريفية من ...   
976334                                                NaN   
672691  Les offres doivent parvenir sous plis fermés p...   
34084   تمكنت فرق الإطفاء التابعة للإدارة الجهوية للحم...   
381147  Après une carrière militaire florissante, le c...   

        Meta Description 1 Length  Meta Description 1 Pixel Width  ...  \
203668                        NaN                             NaN  ...   
280599                      153.0                           833.0  ...   
349421                        NaN                             NaN  ...   
24971                        29.0                           169.0  ...   
342780                      152.0                           835.0  ...   
191977                      153.0                           836.0  ...   
976334                        NaN                             NaN  ...   
672691                      514.0                          3158.0  ...   
34084                       153.0                           840.0  ...   
381147                      153.0                           897.0  ...   

       H2-1 Length  Meta Robots 1 Size (bytes)  Word Count Sentence Count  \
203668         NaN            NaN         3766          38            6.0   
280599        50.0            NaN        39958          76           16.0   
349421         NaN            

In [ ]:
hp_html.dropna(subset='Title 1', inplace=True)

In [ ]:
hp_html = hp_html[~hp_html['Title 1'].str.contains('الساعة')]

In [ ]:
hp_html.shape

(85798, 25)

In [ ]:
# After trying many links with H1 = NaN we found most of them are irrelevant and must be removed except for orientini
hp_html = hp_html[~(hp_html['H1-1'].isna() & ~(hp_html['Address'].str.contains('orientini')))]

In [ ]:
hp_html.shape

(70273, 25)

In [ ]:
hp_html.isna().sum()

Address                               0
Status                                0
Indexability                          0
Indexability Status               70273
Title 1                               0
Title 1 Length                        0
Title 1 Pixel Width                   0
Meta Description 1                16585
Meta Description 1 Length         16585
Meta Description 1 Pixel Width    16585
Meta Keywords 1                   41973
Meta Keywords 1 Length            41973
H1-1                               8712
H1-1 Length                        8712
H2-1                               7296
H2-1 Length                        7296
Meta Robots 1                     64518
Size (bytes)                          0
Word Count                            0
Sentence Count                        0
Average Words Per Sentence            0
Flesch Reading Ease Score             0
Readability                           0
Text Ratio                            0
Last Modified                     48944


In [ ]:
hp_html = hp_html[~hp_html['H2-1'].isna()]

In [ ]:
hp_html.shape

(62977, 25)

In [ ]:
# We found out that links with size < 50000 are irrelevant
hp_html[hp_html['Size (bytes)'] < 50000].sample(5)

Address Status  \
1268597  https://www.interieur.gov.tn/actualite/3123/ال...     OK   
434675   https://www.transtu.tn/fr/achats-publiques/053...     OK   
776290   https://www.interieur.gov.tn/actualite/7499/ال...     OK   
1050952  https://www.interieur.gov.tn/actualite/21541/أ...     OK   
152728   https://www.interieur.gov.tn/actualite/11253/ك...     OK   

        Indexability Indexability Status  \
1268597    Indexable                 NaN   
434675     Indexable                 NaN   
776290     Indexable                 NaN   
1050952    Indexable                 NaN   
152728     Indexable                 NaN   

                                                   Title 1  Title 1 Length  \
1268597  القصرين – سيدي بوزيد - تطاوين / ضبط بضائع مهرب...            79.0   
434675                                             Transtu             7.0   
776290   الجم - المهدية / حجز بضائع مهربة مختلفة على مت...            67.0   
1050952   أهمّ تدخّلات الحماية المدنيّة ليوم 22 جانفي 2021            48.0   
152728   كاب زبيب- بنزرت/ القبض على 07 أشخاص كانوا بصدد...            73.0   

         Title 1 Pixel Width                            Meta Description 1  \
1268597                657.0  Portail du Ministère de l'Intérieur Tunisien   
434675                  69.0                                           NaN   
776290                 551.0  Portail du Ministère de l'Intérieur Tunisien   
1050952                416.0  Portail du Ministère de l'Intérieur Tunisien   
152728                 609.0  Portail du Ministère de l'Intérieur Tunisien   

         Meta Description 1 Length  Meta Description 1 Pixel Width  ...  \
1268597                       44.0                           253.0  ...   
434675                         NaN                             NaN  ...   
776290                        44.0                           253.0  ...   
1050952                       44.0                           253.0  ...   
152728                        44.0                           253.0  ...   

        H2-1 Length  Meta Robots 1 Size (bytes)  Word Count Sentence Count  \
1268597         8.0            NaN        46943         331           96.0   
434675        101.0            NaN        34165         770          117.0   
776290          8.0            NaN        46099         352           99.0   
1050952         8.0            NaN        43729         315           97.0   
152728          8.0            NaN        46898         335           97.0   

         Average Words Per Sentence Flesch Reading Ease Score  Readability  \
1268597                       3.448                   100.000    Very Easy   
434675                        6.581                    79.188  Fairly Easy   
776290                        3.556                   100.000    Very Easy   
1050952                       3.247                   100.000    Very Easy   
152728                        3.454                   100.000    Very Easy   

         Text Ratio  Last Modified  
1268597       6.947            NaT  
434675       15.381            NaT  
776290        7.372            NaT  
1050952       7.269            NaT  
152728        7.065            NaT  

[5 rows x 25 columns]

In [ ]:
hp_html = hp_html[~(hp_html['Size (bytes)'] < 50000)]

In [ ]:
hp_html.shape

(44329, 25)

In [ ]:
hp_html.head()

Address Status Indexability  \
6   https://www.atct.tn/fr/arabie-saoudite-recrute...     OK    Indexable   
15  https://www.atct.tn/fr/node/80573/node/ar/node...     OK    Indexable   
39  https://www.interieur.gov.tn/filtre/categorie/...     OK    Indexable   
53  https://www.defense.tn/category/actualite/actu...     OK    Indexable   
75  https://www.jat.tn/ar/publicationjurisprudence...     OK    Indexable   

   Indexability Status                                            Title 1  \
6                  NaN  Arabie saoudite : recrutement de médecins spéc...   
15                 NaN  How do you judge the access to the spaces on t...   
39                 NaN       Portail du Ministère de l'Intérieur Tunisien   
53                 NaN             Actualités du ministère – Page 2 – MDN   
75                 NaN  القرار التعقيبي الصادر في القضية عدد 310667 بت...   

    Title 1 Length  Title 1 Pixel Width  \
6             71.0                633.0   
15            61.0                541.0   
39            44.0                371.0   
53            38.0                359.0   
75            92.0                804.0   

                                   Meta Description 1  \
6   Le Ministère de la Santé d’Arabie Saoudite dés...   
15                                                NaN   
39       Portail du Ministère de l'Intérieur Tunisien   
53                                                NaN   
75                                              تحميل   

    Meta Description 1 Length  Meta Description 1 Pixel Width  ...  \
6                       221.0                          1287.0  ...   
15                        NaN                             NaN  ...   
39                       44.0                           253.0  ...   
53                        NaN                             NaN  ...   
75                        5.0                            30.0  ...   

   H2-1 Length  Meta Robots 1 Size (bytes)  Word Count Sentence Count  \
6          8.0            NaN        91173         634          173.0   
15         8.0            NaN        87687         551          159.0   
39         6.0            NaN        93693        1978          456.0   
53        70.0            NaN       205951        1109          243.0   
75        20.0            NaN        64240         357          108.0   

    Average Words Per Sentence Flesch Reading Ease Score  Readability  \
6                        3.665                    93.162    Very Easy   
15                       3.465                    95.073    Very Easy   
39                       4.338                   100.000    Very Easy   
53                       4.564                    46.429         Hard   
75                       3.306                   100.000    Very Easy   

    Text Ratio             Last Modified  
6        6.162                       NaT  
15       5.799                       NaT  
39      14.325                       NaT  
53       9.066                       NaT  
75       3.797 2023-07-28 14:27:50+00:00  

[5 rows x 25 columns]

In [ ]:
hp_html = hp_html[~(hp_html['Address'].str.contains('atct') & hp_html['Address'].str.contains('isdb'))]

In [ ]:
hp_html.sample(5)

Address Status Indexability  \
496831  https://orientini.com/FR/Annonces_Orientation_...     OK    Indexable   
694666  http://www.education.gov.tn/?m=202205&paged=2&...     OK    Indexable   
116996  https://www.jat.tn/ar/publicationjurisprudence...     OK    Indexable   
281570  https://orientini.com/FR/Annonces_Orientation_...     OK    Indexable   
603635  https://www.jat.tn/ar/publicationjurisprudence...     OK    Indexable   

       Indexability Status                                            Title 1  \
496831                 NaN  المركز القطاعي للتكوين والتقنيات التطبيقية في ...   
694666                 NaN  mai 2022 – Page 2 – Ministère de l'éducation T...   
116996                 NaN  الحكم الإبتدائي الصادر في القضية عدد 11563 بتا...   
281570                 NaN  جامعة قابس : قائمة الطلبة المقبولين للحصول على...   
603635                 NaN  الحكم الإبتدائي الصادر في القضية عدد 122270 بت...   

        Title 1 Length  Title 1 Pixel Width  \
496831           122.0                992.0   
694666            52.0                473.0   
116996            90.0                763.0   
281570            92.0                747.0   
603635            91.0                774.0   

                                       Meta Description 1  \
496831  Guide interactif de l'orientation, filières un...   
694666                                                NaN   
116996                                              تحميل   
281570  Guide interactif de l'orientation, filières un...   
603635                                              تحميل   

        Meta Description 1 Length  Meta Description 1 Pixel Width  ...  \
496831                      194.0                          1134.0  ...   
694666                        NaN                             NaN  ...   
116996                        5.0                            30.0  ...   
281570                      194.0                          1134.0  ...   
603635                        5.0                            30.0  ...   

       H2-1 Length  Meta Robots 1 Size (bytes)  Word Count Sentence Count  \
496831        92.0            NaN       285240        2586          492.0   
694666        26.0            NaN        79028         734          199.0   
116996        20.0            NaN        66970         420          113.0   
281570        63.0            NaN       312003        2763          536.0   
603635        20.0            NaN        64675         364          108.0   

        Average Words Per Sentence Flesch Reading Ease Score  Readability  \
496831                       5.256                     100.0    Very Easy   
694666                       3.688                     100.0    Very Easy   
116996                       3.717                     100.0    Very Easy   
281570                       5.155                     100.0    Very Easy   
603635                       3.370                     100.0    Very Easy   

        Text Ratio             Last Modified  
496831       8.649                       NaT  
694666       9.484                       NaT  
116996       4.354 2023-07-28 15:46:48+00:00  
281570       8.258                       NaT  
603635       3.934 2023-07-28 15:58:20+00:00  

[5 rows x 25 columns]

## Low-Priority HTML

In [ ]:
lp_html = html_no_dup.copy()

# Drop high-priority HTML rows from lp_html
lp_html = lp_html.drop(hp_html.index)

In [ ]:
lp_html.shape

(71651, 25)

In [ ]:
# Reset the index of both datasets
lp_html.reset_index(drop=True, inplace=True)

In [ ]:
# Reset the index of both datasets
hp_html.reset_index(drop=True, inplace=True)

In [ ]:
pdf.reset_index(drop=True, inplace=True)

In [ ]:
lp_html.index

RangeIndex(start=0, stop=71651, step=1)

In [ ]:
hp_html.index

RangeIndex(start=0, stop=44323, step=1)

In [ ]:
pdf.to_csv('pdf.csv', sep=',', encoding='utf-8')
hp_html.to_csv('hp_html.csv', sep=',', encoding='utf-8')
lp_html.to_csv('lp_html.csv', sep=',', encoding='utf-8')